patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java

```diff
@@ -0,0 +1,239 @@
+		Type annotationType = Type.getType(entry.annotationType);
+		Type targetInType = Type.getType("L" + entry.targetInClass + ";");
+		return new AnnotationData(
+				annotationType, entry.targetType, targetInType, entry.target
+		);
+	}
+
+	public static class AnnotationData {
+		private final Type annotationType;
+		private final ElementType targetType;
+		private final Type clazz;
+		private final String memberName;
+
+		//lazy evaluated
+		private Map<String, Object> annotationData;
+
+		public AnnotationData(
+				final Type annotationType, final ElementType targetType,
+				final Type clazz, final String memberName
+		) {
+			this.annotationType = annotationType;
+			this.targetType = targetType;
+			this.clazz = clazz;
+			this.memberName = memberName;
+		}
+
+		public Type getAnnotationType() {
+			return annotationType;
+		}
+
+		public ElementType getTargetType() {
+			return targetType;
+		}
+
+		public Type getClassType() {
+			return clazz;
+		}
+
+		public String getMemberName() {
+			return memberName;
+		}
+
+		public Map<String, Object> getAnnotationData() {
+			if (annotationData == null) {
+				initAnnotationData();
+			}
+
+			return annotationData;
+		}
+
+		private void initAnnotationData() {
+			annotationData = new HashMap<>();
+
+			try {
+				// TODO: This *may* load classes in the wrong order, but it shouldn't be an issue
+				Class<?> clazzObj = Class.forName(clazz.getClassName());
+				Class annotationType = Class.forName(this.annotationType.getClassName());
+				Annotation annotationObject = getAnnotationObject(clazzObj, annotationType);
+
+				if (annotationObject == null) {
+					LOGGER.error(String.format("Cannot fetch annotation object %s %s %s %s",
+							annotationType, targetType, clazz, memberName
+					));
+					return;
+				}
+
+				Method[] argMethods = annotationObject.getClass().getDeclaredMethods();
+
+				for (Method argMethod : argMethods) {
+					if (isArgumentMethod(argMethod)) {
+						annotationData.put(
+								argMethod.getName(),
+								argMethod.invoke(annotationObject)
+						);
+					}
+				}
+			} catch (Throwable e) {
+				LOGGER.catching(e);
+			}
+		}
+
+		private static boolean isArgumentMethod(Method method) {
+			String name = method.getName();
+			if (name.equals("toString")) return false;
+			if (name.equals("hashCode")) return false;
+			if (name.equals("getClass")) return false;
+			if (name.equals("equals")) return false;
+			if (name.equals("annotationType")) return false;
+
+			return true;
+		}
+
+		private Annotation getAnnotationObject(Class<?> clazzObj, Class annotationType) throws Throwable {
+			switch (targetType) {
+			case TYPE:
+				return clazzObj.getAnnotation(annotationType);
+			case FIELD:
+				return clazzObj.getField(memberName)
+						.getAnnotation(annotationType);
+			case METHOD:
+				//TODO handle overloaded methods
+				String methodName = memberName.substring(0, memberName.indexOf('('));
+				return Arrays.stream(clazzObj.getDeclaredMethods())
```

Comment:
This TODO should either be solved or alternatively assert that the stream's iterator size is 1 after the filter.